### PR TITLE
Add precondition message for Tensor initialization

### PR DESCRIPTION
**Pull request by jon-tow**
Adds precondition messages for failing tensor initializer, `init(shape:scalars:)`, and failing scalarized, `scalarized()`.

### CODE REVIEW

File: Sources/TensorFlow/Core/Tensor.swift [hunk 1]
```diff
@@ -229,6 +230,8 @@ public extension Tensor {
     /// - Precondition: The number of scalars must equal the product of the dimensions of the shape.
     @inlinable
     init(shape: TensorShape, scalars: [Scalar]) {
+        precondition(scalars.count == shape.contiguousSize,
+            "The number of scalars does not equal the product of dimensions given in the shape.")
```
Comment (saeta) [inline Sources/TensorFlow/Core/Tensor.swift]: Could you include the `scalars.count` and `shape.contiguousSize` in the error messages here and below?

Comment (jon-tow) [reply Sources/TensorFlow/Core/Tensor.swift]: That's a great idea. I'll add those in!

Comment (jon-tow) [reply Sources/TensorFlow/Core/Tensor.swift]: I also reordered the precondition values to match the order of the argument labels in the function definition.

File: Sources/TensorFlow/Core/Tensor.swift [hunk 2]
```diff
@@ -239,11 +246,14 @@ public extension Tensor {
...
     ///   - scalars: The scalar contents of the tensor.
-    /// - Precondition: The number of scalars must equal the product of the
-    ///   dimensions of the shape.
+    /// - Precondition: The product of the dimensions of the shape must equal the number of scalars.
     @inlinable
     init(shape: TensorShape, scalars: UnsafeBufferPointer<Scalar>) {
-        precondition(scalars.count == shape.contiguousSize)
+        precondition(shape.contiguousSize == scalars.count,
+            """
+            The product of the dimensions of the shape must equal the number of scalars:
```
Comment (rxwei) [inline Sources/TensorFlow/Core/Tensor.swift]: To make this message be one line.
```suggestion
            The product of the dimensions of the shape must equal the number of scalars: \
```

File: Sources/TensorFlow/Core/Tensor.swift [hunk 3]
```diff
@@ -226,9 +228,14 @@ public extension Tensor {
     /// - Parameters:
     ///   - shape: The shape of the tensor.
     ///   - scalars: The scalar contents of the tensor.
-    /// - Precondition: The number of scalars must equal the product of the dimensions of the shape.
+    /// - Precondition: The product of the dimensions of the shape must equal the number of scalars.
     @inlinable
     init(shape: TensorShape, scalars: [Scalar]) {
+        precondition(shape.contiguousSize == scalars.count,
+            """
+            The product of the dimensions of the shape must equal the number of scalars:
```
Comment (rxwei) [inline Sources/TensorFlow/Core/Tensor.swift]:
```suggestion
            The product of the dimensions of the shape must equal the number of scalars: \
```

File: Sources/TensorFlow/Core/Tensor.swift [hunk 4]
```diff
@@ -140,6 +140,8 @@ public extension Tensor {
     @inlinable
     @differentiable(wrt: self, vjp: _vjpScalarized where Scalar: TensorFlowFloatingPoint)
     func scalarized() -> Scalar {
+        precondition(shape.contiguousSize == 1,
+           "This tensor must have exactly one scalar: \(shape.contiguousSize).")
```
Comment (rxwei) [inline Sources/TensorFlow/Core/Tensor.swift]:
```suggestion
           "This tensor must have exactly one scalar to be scalarized, but it has \(shape.contiguousSize) scalars.")
```

File: Sources/TensorFlow/Core/Tensor.swift [hunk 5]
```diff
@@ -257,11 +267,14 @@ public extension Tensor {
...
     ///   - scalars: The scalar contents of the tensor.
-    /// - Precondition: The number of scalars must equal the product of the
-    ///   dimensions of the shape.
+    /// - Precondition: The product of the dimensions of the shape must equal the number of scalars.
     @inlinable
     init<C: RandomAccessCollection>(shape: TensorShape, scalars: C) where C.Element == Scalar {
-        precondition(scalars.count == shape.contiguousSize)
+        precondition(shape.contiguousSize == scalars.count,
+            """
+            The product of the dimensions of the shape must equal the number of scalars:
```
Comment (rxwei) [inline Sources/TensorFlow/Core/Tensor.swift]:
```suggestion
            The product of the dimensions of the shape must equal the number of scalars: \
```

File: Sources/TensorFlow/Core/Tensor.swift [hunk 6]
```diff
@@ -226,9 +228,14 @@ public extension Tensor {
...
     ///   - shape: The shape of the tensor.
     ///   - scalars: The scalar contents of the tensor.
-    /// - Precondition: The number of scalars must equal the product of the dimensions of the shape.
+    /// - Precondition: The product of the dimensions of the shape must equal the number of scalars.
     @inlinable
     init(shape: TensorShape, scalars: [Scalar]) {
+        precondition(shape.contiguousSize == scalars.count,
+            """
+            The product of the dimensions of the shape must equal the number of scalars: \
+            \(shape.contiguousSize) vs. \(scalars.count)
```
Comment (saeta) [inline Sources/TensorFlow/Core/Tensor.swift]: This is looking really great. One last thought: what do you think of an error message like the following?

```suggestion
            the shape requires \(shape.contiguousSize) scalars but only \(scalars.count) were provided.
```

File: Sources/TensorFlow/Core/Tensor.swift [hunk 7]
```diff
@@ -140,6 +140,8 @@ public extension Tensor {
     @inlinable
     @differentiable(wrt: self, vjp: _vjpScalarized where Scalar: TensorFlowFloatingPoint)
     func scalarized() -> Scalar {
+        precondition(shape.contiguousSize == 1,
+           "This tensor must have exactly one scalar to be scalarized, but it has \(shape.contiguousSize) scalars.")
```
[the remark this answers is not in the document]
Comment (jon-tow) [reply Sources/TensorFlow/Core/Tensor.swift]: Sure thing! The below change has less than 100 columns. How does it sound to you?
"This tensor must have exactly one scalar but contains \(shape.contiguousSize)."

[the remark this answers is not in the document]
Comment (rxwei) [reply Sources/TensorFlow/Core/Tensor.swift]: Sounds good!

File: Sources/TensorFlow/Core/Tensor.swift (hunk 6 above)
Comment (jon-tow) [reply Sources/TensorFlow/Core/Tensor.swift]: I do like this message better. The mapping from arguments to their problematic values is clearer. I'll update them!

**Pull request merged**